renderdoc/driver/metal/metal_device.cpp

```diff
@@ -38,3 +51,121 @@ MTL::Device *WrappedMTLDevice::MTLCreateSystemDefaultDevice(MTL::Device *realMTL
 
   return UnwrapObjC<MTL::Device *>(wrappedMTLDevice);
 }
+
+MTL::Library *WrappedMTLDevice::newDefaultLibrary()
+{
+  MTL::Library *realMTLLibrary;
+
+  SERIALISE_TIME_CALL(realMTLLibrary = Unwrap<MTL::Device *>(this)->newDefaultLibrary());
+  MetalResourceManager::UnwrapHelper<MTL::Library *>::Outer *wrappedMTLLibrary;
+  ResourceId id = GetResourceManager()->WrapResource(realMTLLibrary, wrappedMTLLibrary);
+  if(IsCaptureMode(m_State))
+  {
+    void *pData;
+    uint32_t bytesCount;
+    ObjC::Get_defaultLibraryData(pData, bytesCount);
+    Chunk *chunk = NULL;
+    {
+      CACHE_THREAD_SERIALISER();
+      SCOPED_SERIALISE_CHUNK(MetalChunk::MTLDevice_newDefaultLibrary);
+      Serialise_mtlDevice_newDefaultLibrary(ser, this, wrappedMTLLibrary, pData, bytesCount);
+      chunk = scope.Get();
+    }
+    free(pData);
+    MetalResourceRecord *record = GetResourceManager()->AddResourceRecord(wrappedMTLLibrary);
+    record->AddChunk(chunk);
+    GetResourceManager()->MarkResourceFrameReferenced(id, eFrameRef_Read);
+  }
+  else
+  {
+    // TODO: implement RD MTL replay
+    //     GetResourceManager()->AddLiveResource(id, wrappedMTLLibrary);
+  }
+  return UnwrapObjC<MTL::Library *>(wrappedMTLLibrary);
+}
+
+MTL::Library *WrappedMTLDevice::newLibraryWithSource(NS::String *source,
+                                                     MTL::CompileOptions *options, NS::Error **error)
+{
+  MTL::Library *realMTLLibrary;
+  SERIALISE_TIME_CALL(realMTLLibrary =
+                          Unwrap<MTL::Device *>(this)->newLibrary(source, options, error));
+  MetalResourceManager::UnwrapHelper<MTL::Library *>::Outer *wrappedMTLLibrary;
+  ResourceId id = GetResourceManager()->WrapResource(realMTLLibrary, wrappedMTLLibrary);
+  if(IsCaptureMode(m_State))
+  {
+    Chunk *chunk = NULL;
+    {
+      CACHE_THREAD_SERIALISER();
+      SCOPED_SERIALISE_CHUNK(MetalChunk::MTLDevice_newLibraryWithSource);
+      Serialise_mtlDevice_newLibraryWithSource(ser, this, wrappedMTLLibrary, source, options);
+      chunk = scope.Get();
+    }
+    MetalResourceRecord *record = GetResourceManager()->AddResourceRecord(wrappedMTLLibrary);
+    record->AddChunk(chunk);
+    GetResourceManager()->MarkResourceFrameReferenced(id, eFrameRef_Read);
+  }
+  else
+  {
+    // TODO: implement RD MTL replay
+    //     GetResourceManager()->AddLiveResource(id, wrappedMTLLibrary);
+  }
+  return UnwrapObjC<MTL::Library *>(wrappedMTLLibrary);
+}
+
+template <typename SerialiserType>
+bool WrappedMTLDevice::Serialise_mtlDevice_newDefaultLibrary(SerialiserType &ser,
```

Comment:
By convention I always organise functions with the `Serialise_` version immediately before the real implementation. No particular reason, but it would be better to follow that convention.

Similarly, the `mtlDevice` is redundant in this function name because this is already implemented in `WrappedMTLDevice` so it should be `WrappedMTLDevice::Serialise_newDefaultLibrary`.

In general it's preferred to have serialise functions have the same signature as the function they are serialising, unless there's a strong reason to deviate. Cases like this where the return value needs to be serialised in some way would be one case, but in general where some parameters are implicit instead of explicit and need to be fetched like the pData here, it would be preferable to fetch them inside this function ahead of the serialisation in an `if(ser.IsWriting())` block, unless they are only fetchable in the outside real wrapper function.

I'm unclear why the `device` parameter is here. It seems like that is always going to be `this`? For the device where we only have one, it can be omitted entirely, for other functions where you need to serialise the current object you can serialise `this`.

Reply:
Now the ObjC helper APIs are not needed and call directly to the real APIs from C++. This can be updated and follow existing patterns using helper macros like this

#define IMPLEMENT_FUNCTION_SERIALISED(ret, func, ...) \
  ret func(__VA_ARGS__);                              \
  template <typename SerialiserType>                  \
  bool CONCAT(Serialise_, func(SerialiserType &ser, __VA_ARGS__));

#define INSTANTIATE_FUNCTION_SERIALISED(CLASS, ret, func, ...)                                    \
  template bool CLASS::CONCAT(Serialise_, func(ReadSerialiser &ser, __VA_ARGS__)); \
  template bool CLASS::CONCAT(Serialise_, func(WriteSerialiser &ser, __VA_ARGS__));

The intent of the naming was to have the serialization method name match the chunk name and because the APIs are OO and not global wanted to include the name of the parent object in the chunk name just in case there is an API name that is the same in two different objects (say perhaps a common API name like "update"). Will keep the chunk name being unique and update the serialization macros and usage to match more existing patterns and remove unused parameters

Reply:
Yes the chunk can have the object namespace e.g. `MetalChunk::Device_newDefaultLibrary` or similar, but it can be omitted from the function name itself as the namespacing already exists in the definition. Good to know this can be refactored then :+1:.

Reply:
1f9a217577fe36c8bc218da0bbb2f3e721f22011